### PR TITLE
Add Grana's connected components algorithm for 8-way connectivity. Th…

### PR DESCRIPTION
**Pull request by prittt**
### What does this PR change?
<p align="justify">This PR introduces a faster algorithm for Connected Components Labeling with 8-way connectivity. The algorithm has been originally proposed in:
</br></br>
Grana, Costantino, Daniele Borghesani, and Rita Cucchiara. "Optimized block-based connected components labeling with decision trees." Image Processing, IEEE Transactions on 19.6 (2010): 1596-1609.
</br></br>
A series of performance comparisons has been made to assess the execution times of this algorithm, and to compare its performance with respect to the the algorithm currently implemented in OpenCV. Results of these evaluations, along with their source code, can be found in the <a href="https://github.com/prittt/YACCLAB">YACCLAB</a> project on Github.
The algorithm this PR introduces is named as BBDT in that project, while the current CCL algorithm implemented in Opencv is called SAUF.</p>

### Notes on *distanceTransformWithLabels* function
<p align="justify"><i>distanceTransformWithLabels</i> is a variant of the <i>distanceTransform</i> function included in Opencv. This variant does not only compute the minimum distance for each pixel (x,y), like <i>distanceTransform</i> does, but also identifies the nearest connected components consisting of zero pixels calling the <i>connectedComponents</i> function (which is the object of this PR). We change <i>connectedComponents</i> calls from <i>distanceTransform</i> function to forces SAUF algorithm execution for the following reasons: </p>
- *distanceTransform* algorithm requires a *labels*'s matrix with row major ordering of labels;
- SAUF algorithm forces that order of labels while BBDT does not.

<p align="justify">It is important to underline that different algorithms can produce the same connected components with different labels, this has not be considered as an error. For example, given the following binary image<p>

0 1 0 0 0 0 1
0 1 1 1 0 0 0
0 0 0 0 0 1 1
1 1 1 1 1 1 1

two different algorithms may label connected components in the following different ways:

0 1 0 0 0 0 2
0 1 1 1 0 0 0
0 0 0 0 0 3 3
3 3 3 3 3 3 3

0 2 0 0 0 0 3
0 2 2 2 0 0 0
0 0 0 0 0 1 1
1 1 1 1 1 1 1

This won't impact on any further analysis or on the statistics computed by the algorithms.

### Contributors
Federico Bolelli,
Costantino Grana,
Lorenzo Baraldi,
Roberto Vezzani.

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
To ensure binary and source compatibility (ABI) you cannot change functions like you are doing. One way around this is to create a wrapper function around the existing function or create a new function completely.

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 1]
```diff
@@ -3595,15 +3602,18 @@ CV_EXPORTS_W void matchTemplate( InputArray image, InputArray templ,
...
 @param image the 8-bit single-channel image to be labeled
 @param labels destination labeled image
 @param connectivity 8 or 4 for 8-way or 4-way connectivity respectively
 @param ltype output image label type. Currently CV_32S and CV_16U are supported.
- */
+@param ccltype connected components algorithm type (see the cv::ConnectedComponentsAlgorithmsTypes).
+*/
 CV_EXPORTS_W int connectedComponents(InputArray image, OutputArray labels,
-                                     int connectivity = 8, int ltype = CV_32S);
+                                     int connectivity = 8, int ltype = CV_32S, int ccltype = CCL_DEFAULT);
```
Comment (StevenPuttemans) [inline modules/imgproc/include/opencv2/imgproc.hpp]: This is what is breaking binary compatibility and source compatibility by changing the number of parameters.
One way could be to add a new function `connectedComponentsGrana`

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 2]
```diff
@@ -3615,11 +3625,11 @@ cv::ConnectedComponentsTypes. The data type is CV_32S.
 accessed via centroids(label, 0) for x and centroids(label, 1) for y. The data type CV_64F.
 @param connectivity 8 or 4 for 8-way or 4-way connectivity respectively
 @param ltype output image label type. Currently CV_32S and CV_16U are supported.
+@param ccltype connected components algorithm type (see the cv::ConnectedComponentsAlgorithmsTypes).
 */
 CV_EXPORTS_W int connectedComponentsWithStats(InputArray image, OutputArray labels,
                                               OutputArray stats, OutputArray centroids,
-                                              int connectivity = 8, int ltype = CV_32S);
-
+                                              int connectivity = 8, int ltype = CV_32S, int ccltype = CCL_DEFAULT);
```
Comment (StevenPuttemans) [inline modules/imgproc/include/opencv2/imgproc.hpp]: Same problem here.

**Comment by vpisarev** [comment]
this PR will be recreated because of CI issues

**Pull request closed without merging**